python/graphstorm/wholegraph/wholegraph.py

```
@@ -283,3 +292,336 @@ def load_wg_feat(part_config_path, num_parts, type_name, name):
+        If so, defer the creation of WG tensor until the WG optimizer is created/attached.
+    """
+    def __init__(
+        self,
+        shape: tuple,
+        dtype: th.dtype,
+        name: str,
+        location: str = "cpu",
+        use_wg_optimizer: bool = False,
+    ):
+        self._nnodes = shape[0]
+        self._embedding_dim = shape[1]
+        self._name = name
+        self._dtype = dtype
+        self._location = location
+        self._use_wg_optimizer = use_wg_optimizer
+        # Need the pylibwholegraph be at least 23.12.00 to support _tensor.scatter API.
+        assert pylibwholegraph.__version__ >= "23.12.00", \
+            "Please upgrade to the latest version of WholeGraph."
+        self._tensor = None
+        self._module = None
+        self._optimizer = None
+        # When _use_wg_optimizer, we have _module -> _tensor -> optimizer (-> means "depends on")
+        # So, optimizer has to be created first, then tensor, then module.
+        if not self._use_wg_optimizer:
+            # Otherwise, we can initialize _tensor here
+            self._tensor = create_wg_dist_tensor(shape, dtype, location)
+
+    def attach_wg_optimizer(self, wg_optimizer=None):
+        """
+        Attach a WholeGraph sparse optimizer to the WholeGraph embedding.
+        This is needed for trainable embeddings
+
+        Parameters
+        ----------
+        wg_optimizer : WholeMemoryOptimizer
+            The WholeGraph sparse optimizer to be attached to the WholeGraph embedding.
+
+        Returns
+        -------
+        None
+
+        """
+        assert self._use_wg_optimizer, \
+            "Please create WholeGraphDistTensor tensor with use_wg_optimizer=True."
+        if self.optimizer == wg_optimizer and self._module is not None:
+            # no-op if the optimizer is the same
+            return
+        # When attach a new optimizer, we have to purge the old _tensor/_module/_optimizer.
```

If self.optimizer is not None, Can we just raise an error here?
Ideally, we do not want to initialize the optimizer twice.